### PR TITLE
 Add support for R_X86_64_TPOFF32 and R_X86_64_TPOFF64 relocations

### LINKED ISSUE #534: Implement x86_64 target backend in ELD
Summary
- Implement an x86_64 target backend for the ELD linker.
- Phase 1 targets static linking support (Linux + baremetal), with the goal of passing unit tests and producing working binaries.
- Future phases can extend to dynamic linking, symbol versioning, and broader GNU compatibility.

Goals
- Add a new x86_64 target backend that:
  - Builds in the ELD tree and integrates with ELD’s plugin/target architecture
  - Handles core ELF x86_64 relocations for static linking, including common PC-relative and absolute relocations
  - Supports basic TLS relocations needed for static/PIC cases
  - Produces detailed map files
  - Works with ELD’s linker script parser (GNU-compatible where applicable)
  - Normalizes target triples (e.g., x86_64-linux-gnu vs x86_64-unknown-linux-gnu)
  - Passes CI on Linux; optional Windows build check if feasible

Non-Goals (Phase 1)
- Full dynamic linking (GLIBC symbol versioning, PLT/GOT for shared libraries)
- Advanced GNU script edge cases beyond commonly used constructs
- Comprehensive Windows support (can be planned separately)

Scope & Deliverables
- New x86_64 target backend module(s) in ELD
- Core relocation handling (examples: 64-bit absolute, 32/32S, common PC-relative forms, minimal TLS)
- Section merging (.text, .data, .bss), alignment, start address handling aligned with triple/OS
- Linker script compatibility for common patterns; ensure parser handles “: ALIGN(…)” spacing expectations
- Detailed map-file output with target triple environment
- Reproduce tar functionality remains working when linking x86_64
- Docs: brief “Getting started” + “Supported features/limitations” + “Examples”
- Tests: LIT/unit tests for relocations, script handling, map output

User Scenarios
- Static link a simple C/C++ program targeting x86_64 (baremetal and Linux/musl)
- Link with a simple GNU-style linker script (SECTIONS, ALIGN, basic placement)
- Produce a readable map file confirming sections, symbols, and target triple environment

Work Breakdown (Checklist)
1) Environment & Planning
   - [x] Build ELD locally (Linux) and read target backend guidelines
   - [x] Confirm initial scope (static linking, minimal TLS, map files, scripts)
   - [ ] Pick initial triples to support (e.g., x86_64-unknown-linux-gnu, x86_64-none-elf)

2) Target Backend Skeleton
   - [x] Create x86_64 target descriptor (endianness, pointer size, data layout)
   - [ ] Register backend and wiring in CMake/build

3) Object & Sections
   - [x] Parse ELF x86_64 objects and collect sections/symbols
   - [ ] Implement section merging (.text/.data/.bss), alignment rules
   - [x] Ensure init_array/fini_array support in line with ELD direction

4) Relocations (Phase 1 set)
   - [x] Implement core absolute and PC-relative relocations commonly needed for static linking
   - [x] Minimal TLS models sufficient for static/PIC cases (document exactly what’s supported)
   - [ ] Add relocation-specific diagnostics

5) Linker Script Compatibility
   - [x] Validate common GNU constructs (SECTIONS, ALIGN, KEEP, GROUP, etc.)
   - [ ] Ensure parser behavior with “: ALIGN(…)” vs “:ALIGN(…)”; document user guidance

6) Triple Handling & Defaults
   - [ ] Normalize x86_64 triples and map OS environment for map-file output
   - [ ] Ensure consistent behavior when LLVM default triple differs (unknown vs linux)

7) Map Files & Reproduce
   - [ ] Add x86_64 metadata to detailed map file (environment/triple, layout)
   - [ ] Verify reproduce tar generation works on failures

8) CI & Tooling
   - [ ] Add GitHub Actions workflow (Linux) building ELD + running tests
   - [ ] Optional: Windows workflow stub (tracked separately if needed)

9) Tests
   - [ ] LIT tests: relocations (absolute, PC-relative), TLS (if in scope), script parsing (ALIGN), map-file content
   - [ ] Sample baremetal and Linux static link tests
   - [ ] Negative tests for invalid script/input section descriptions (ensure proper errors)

10) Documentation
   - [ ] Add “x86_64 backend – quick start” doc: how to build, invoke, and known limitations
   - [ ] Update linker script GNU-compat docs and examples (include ALIGN spacing note)
   - [ ] Add a simple end-to-end example

Risks & Dependencies
- Triple normalization can change layout assumptions (e.g., marking OS as linux vs none); tests must set explicit triples
- Windows builds may require minor source changes and CI adjustments
- Future dynamic linking/symbol versioning is out-of-scope; plan a follow-up issue

Testing Plan
- Unit/LIT tests for:
  - Core relocations resolve correctly and produce expected bytes
  - Script parsing: SECTIONS with ALIGN; verify no crashes and correct diagnostics
  - Map-file: includes “Target triple environment” and sane layout details
- End-to-end: build+link sample x86_64 static binaries (baremetal and Linux/musl)
- Reproduce tar: ensure it captures inputs for debugging

Definition of Done (Phase 1)
- ✅ ELD builds with the x86_64 backend enabled on Linux
- ✅ Core relocations pass unit/LIT tests
- ✅ A simple static binary links and runs (baremetal or Linux/musl), with correct sections/layout
- ✅ Map-file shows triple environment and section/symbol details
- ✅ CI green on Linux (libc test suite using musl); docs published; known limitations listed

### PR DESCRIPTION
**Pull request by bhoumik03**
- Added `relocTPOFF` function implementing the TPOFF formula: `S + A - TLS_Template_Size`
- Added TLS template size calculation in `computeTLSOffsets()` method
- Fixed type handling

Fixes #534

### CODE REVIEW

**Comment by parth-07** [comment]
Can you please add the test case for relocation overflow with TPOFF32 as we discussed offline?

File: lib/Target/x86_64/x86_64Relocator.cpp [hunk 1]
```diff
@@ -263,6 +266,17 @@ Relocator::Result VerifyRelocAsNeededHelper(
   return R;
 }
 
+void x86_64Relocator::computeTLSOffsets() {
+  ELFSegment *tlsSegment =
+      getTarget().elfSegmentTable().find(llvm::ELF::PT_TLS, llvm::ELF::PF_R, 0);
```
Comment (parth-07) [inline lib/Target/x86_64/x86_64Relocator.cpp]: What happens if there are multiple TLS segments?

File: lib/Target/x86_64/x86_64Relocator.cpp [hunk 2]
```diff
@@ -263,6 +266,17 @@ Relocator::Result VerifyRelocAsNeededHelper(
...
+  if (tlsSegment) {
```
Comment (parth-07) [inline lib/Target/x86_64/x86_64Relocator.cpp]: What if there are no tls segments? Should we report an error?

File: .gitignore [hunk 3]
```diff
@@ -44,7 +44,7 @@ autoconf/autom4te.cache
 # CLion project configuration
 /.idea
 /cmake-build*
-
+.clangd
```
Comment (parth-07) [inline .gitignore]: Remove this change.

**Comment by parth-07** [comment]
> Add support for tls reloctions and alignment fix

Can you please improve the commit message? It is misleading because the commit is not fixing all tls relocations. Alignment fix is too vague. And it does not specify x86-64 anywhere.

File: lib/Target/x86_64/x86_64Relocator.cpp (hunk 2 above)
Comment (bhoumik03) [reply lib/Target/x86_64/x86_64Relocator.cpp]: tls segments are optionally so we should not error out generally. We can add check in the tls relocations to error out if there is a tls relocation but no tls segment; are you referring to that?

**Comment by bhoumik03** [comment]
> Can you please add the test case for relocation overflow with TPOFF32 as we discussed offline?

Added the test case in updated PR

File: lib/Target/x86_64/x86_64Relocator.cpp (hunk 1 above)
Comment (bhoumik03) [reply lib/Target/x86_64/x86_64Relocator.cpp]: We assert out that multiple TLS segments are not supported for x86_64 as we had discussed

File: lib/Target/x86_64/x86_64Relocator.cpp (hunk 2 above)
Comment (bhoumik03) [reply lib/Target/x86_64/x86_64Relocator.cpp]: have added check in the relocTPOFF() that ensures that we have tls segment given we have a TPOFF type relocation

**Comment by parth-07** [comment]
> > Can you please change the C-source files used for tests to follow the existing pattern of how C-source files are named in the codebase?
> 
> I am currently naming the relocation test cases as relocNAME.test, should I change this to the numeric naming like 1.c?

Yes. Please use the names such as 1.c.

File: lib/Target/x86_64/x86_64Relocator.cpp [hunk 4]
```diff
@@ -263,6 +266,24 @@ Relocator::Result VerifyRelocAsNeededHelper(
...
+void x86_64Relocator::computeTLSOffsets() {
+  std::vector<ELFSegment *> tlsSegments =
+      getTarget().elfSegmentTable().getSegments(llvm::ELF::PT_TLS);
+
+  if (tlsSegments.empty()) {
+    return;
+  }
+
+  assert(tlsSegments.size() == 1 &&
+         "Multiple TLS segments not supported in x86_64 backend");
```
Comment (parth-07) [inline lib/Target/x86_64/x86_64Relocator.cpp:278]: Can you please open an issue for this?

File: lib/Target/x86_64/x86_64Relocator.cpp [hunk 5]
```diff
@@ -263,6 +266,24 @@ Relocator::Result VerifyRelocAsNeededHelper(
...
 
+void x86_64Relocator::computeTLSOffsets() {
+  std::vector<ELFSegment *> tlsSegments =
+      getTarget().elfSegmentTable().getSegments(llvm::ELF::PT_TLS);
+
+  if (tlsSegments.empty()) {
+    return;
+  }
+
+  assert(tlsSegments.size() == 1 &&
```
Comment (parth-07) [inline lib/Target/x86_64/x86_64Relocator.cpp]: Please use `ASSERT` macro that is used throughout the eld.

File: test/x86_64/linux/relocTPOFF32_overflow/relocTPOFF32_overflow.test [hunk 6]
```diff
@@ -0,0 +1,14 @@
+#--relocTPOFF32_overflow.test----------Executable--------#
+BEGIN_COMMENT
+# Test R_X86_64_TPOFF32 relocation overflow with large TLS template size.
+# Creates a 3.25GB TLS template which should cause TPOFF32 overflow since
+# TPOFF32 is signed 32-bit with range ±2GB.
+#END_COMMENT
+
+#START_TEST
+RUN: %clang %clangopts -fdata-sections -ffunction-sections -c %p/Inputs/1.c -o %t.o
+RUN: not %link %linkopts -T %p/Inputs/script_overflow.t %t.o -o %t.exe 2>&1 | FileCheck %s
```
Comment (parth-07) [inline test/x86_64/linux/relocTPOFF32_overflow/relocTPOFF32_overflow.test]:
```suggestion
RUN: %not %link %linkopts -T %p/Inputs/script_overflow.t %t.o -o %t.exe 2>&1 | FileCheck %s
```
